### ISSUE TITLE
BUG: Repeat Jobs stop processing

### ISSUE THREAD

**Issue by mbernardini** [OptimalBits/bull#747]
We're noticing an issue that when using repeat jobs, the jobs eventually stop processing with no indication of an error. The reason that the jobs stop is that the next job does not get added to the queue. I ran NODE_DEBUG=bull and added some additional debug statements, so I was able to figure out that `nextRepeatableJob` function is being called but [this check]( https://github.com/OptimalBits/bull/blob/ace16ecd0bdeca03b1c1955281a4ab2095f355f8/lib/repeatable.js#L41) is causing the next job to not get added because `now` is a few milliseconds before `millis` (the time the job is supposed to get executed).

This script reproduces the issue:
```Javascript
const Queue = require('bull');
const prefix = 'test';
const redis = {host: 'localhost', port: 6379};

const cronQueue = new Queue('cronQueue', {redis, prefix});
cronQueue.process('cron', () => {
  console.log(`[${(new Date()).toISOString()}] Job Processing...`);
});
cronQueue.add('cron', {}, {repeat:{cron:'*/10 * * * *'}, jobId: 'cronJob', removeOnComplete: true});
```
Output:
```
[2017-10-20T17:20:00.012Z] Job Processing...
[2017-10-20T17:29:59.987Z] Job Processing...
// Theres no more out at this point because the second job executed prior  to the scheduled time
```

For reference:
Bull version: 3.3.2
Node Version: 6.9.0
Redis Version: 2.8.19 and 3.2.4
I've seen this issue both running Redis and Node on a single machine as well as in our cloud environment (clustered Redis and multiple Node instances)

**Comment by ritter** [comment]
I have this issue as well, and have not yet been able to investigate fully.  It appears the jobs run and do not re-queue for their next interval and never run again until re-added.

**Comment by mbernardini** [comment]
I suspect its due to the job getting run a few milliseconds before its scheduled time.

I have a patch that puts a band-aid over the issue:
```Javascript
    queue.nextRepeatableJob = function(name, data, opts, isRepeat){
      const delay = ms => new Promise(r => setTimeout(r, ms));
      return Queue.prototype.nextRepeatableJob.call(this, name, data, opts, isRepeat)
        .then(job => {
          if(job) return job;
          return delay(1000).then(() => Queue.prototype.nextRepeatableJob.call(this, name, data, opts, isRepeat));
        });
    };
```

But I think the underlying issue is in the code that moves the delayed jobs into waiting (so that they are executed) is moving jobs just a little too early.

**Comment by manast** [comment]
I have to investigate more, but the following code running sinon fakeTimers works well: 
```
var _this = this;
    var nextTick = 10 * ONE_MINUTE + 1;

    var date = new Date('2017-02-02 7:21:42');
    this.clock.tick(date.getTime());
    
    this.timeout(50000);
    queue.process('cron', function() {
      console.error((new Date()).toISOString(), 'Job Processing...');
      _this.clock.tick(nextTick);      
    });
    queue.add('cron', 
      {}, 
      {
        repeat: {
          cron:'*/10 * * * *'
        },
        jobId: 'cronJob', removeOnComplete: true
      }).then(function(){
      _this.clock.tick(nextTick);
    });
```
I am tying to figure out how I could simulate the real scenario that triggers the issue.

**Comment by manast** [comment]
I think I found the issue, actually a copy/paste error: https://github.com/OptimalBits/bull/blob/ace16ecd0bdeca03b1c1955281a4ab2095f355f8/lib/repeatable.js#L38
The bad thing was that the unit tests were passing anyway :(. I have tested your test code after removing that line and it is working for me.

**Comment by mbernardini** [comment]
Thanks for fixing this bug!

**Comment by manast** [comment]
@ritter how can 3.3.2 have broken repeatable jobs and not 3.3.1 when there are no changes related to repeatable jobs between the two? And what is the test case?

**Comment by mbernardini** [comment]
I'm not sure about what @ritter saw but we saw this issue in 3.3.1 and 3.3.2.

I re-tested 3.3.3 and I'm not sure that the issue is fixed in 3.3.3. Running the script I originally posted, I'm seeing the repeated job stop after only a couple of iterations. It looks like the code is still returning in this check without creating the next job:
```JavaScript
      return getMillis.then(function(millis){
        millis = parseInt(millis);
        if(millis > now || !millis){
          // We have already a next repeatable job scheduled.
          // or the job has been removed.
          return;
        }
```

**Comment by ponomarenkotema** [comment]
Hi! Any updates here?
I have got similar problem with 3.4.8 version. Job has stoped being processed after couple of iterations.
It has such simple config 
add({}, {
		repeat: {
			every: 60000 * 30
		}
});
And I do not found any hints in debug output.

**Comment by manast** [comment]
Could this be related to #890 ?

**Comment by ponomarenkotema** [comment]
> Could this be related to #890 ?

It seems no, it case of reconnection new job can be added by processed only after app restarted  or queues reinited. 
In this case next delayed job was not added but it new jobs (added by api logic) still working fine.

**Comment by kishore-uptycs** [comment]
Any update on this?
I'm also facing the same issue! 
Currently running: 3.29.1
It stops processing after running for around 6 hrs!

**Comment by manast** [comment]
@kishore-uptycs this issue is closed so there is nothing to update really. If you have a issue please open a new issue and provide reproducible code.
